lambda/trmc.ml

```diff
@@ -0,0 +1,515 @@
+          );
+        }
+
+  and choice_prim ctx prim primargs loc =
+    begin [@warning "-8"] (* see choice *)
+      match prim with
+      (* The important case is the construction case *)
+      | Pmakeblock (tag, flag, shape) ->
+          choice_makeblock ctx (tag, flag, shape) primargs loc
+
+      (* Some primitives have arguments in tail-position *)
+      | (Pidentity | Popaque) as idop ->
+          let [l1] = primargs in
+          lift ctx [l1] (fun [l1] -> Lprim (idop, [l1], loc))
+      | (Psequand | Psequor) as shortcutop ->
+          let [l1; l2] = primargs in
+          lift ctx [l2]
+            (fun [l2] -> Lprim (shortcutop, [l1; l2], loc))
+
+      (* cases we don't handle yet *)
+      | (Prevapply | Pdirapply) ->
+          failwith "TODO: should have been simplified away already"
+
+      | (Pmakearray _ | Pduparray _) ->
+          failwith "TODO: we don't handle array indices as destinations yet"
+
+      | Pduprecord _ ->
+          failwith "TODO"
+
+      (* in common cases we just Return *)
+      | Pbytes_to_string | Pbytes_of_string
+      | Pgetglobal _ | Psetglobal _
+      | Pfield _ | Pfield_computed
+      | Psetfield _ | Psetfield_computed _
+      | Pfloatfield _ | Psetfloatfield _
+      | Pccall _
+      | Praise _
+      | Pnot
+      | Pnegint | Paddint | Psubint | Pmulint | Pdivint _ | Pmodint _
+      | Pandint | Porint | Pxorint
+      | Plslint | Plsrint | Pasrint
+      | Pintcomp _
+      | Poffsetint _ | Poffsetref _
+      | Pintoffloat | Pfloatofint
+      | Pnegfloat | Pabsfloat
+      | Paddfloat | Psubfloat | Pmulfloat | Pdivfloat
+      | Pfloatcomp _
+      | Pstringlength | Pstringrefu  | Pstringrefs
+      | Pbyteslength | Pbytesrefu | Pbytessetu | Pbytesrefs | Pbytessets
+      | Parraylength _ | Parrayrefu _ | Parraysetu _ | Parrayrefs _ | Parraysets _
+      | Pisint | Pisout
+      | (
+        (* operations returning boxed values could be considered constructions someday *)
+        Pbintofint _ | Pintofbint _
+        | Pcvtbint _
+        | Pnegbint _
+        | Paddbint _ | Psubbint _ | Pmulbint _ | Pdivbint _ | Pmodbint _
+        | Pandbint _ | Porbint _ | Pxorbint _ | Plslbint _ | Plsrbint _ | Pasrbint _
+        | Pbintcomp _
+      )
+      | Pbigarrayref _ | Pbigarrayset _
+      | Pbigarraydim _
+      | Pstring_load_16 _ | Pstring_load_32 _ | Pstring_load_64 _
+      | Pbytes_load_16 _ | Pbytes_load_32 _ | Pbytes_load_64 _
+      | Pbytes_set_16 _ | Pbytes_set_32 _ | Pbytes_set_64 _
+      | Pbigstring_load_16 _ | Pbigstring_load_32 _ | Pbigstring_load_64 _
+      | Pbigstring_set_16 _ | Pbigstring_set_32 _ | Pbigstring_set_64 _ | Pctconst _
+      | Pbswap16
+      | Pbbswap _
+      | Pint_as_pointer
+        ->
+          Return (Lprim (prim, primargs, loc))
+    end
+  in choice ctx t
+
+and traverse ctx = function
+  | Lletrec (bindings, body) ->
+      let ctx, bindings = traverse_letrec ctx bindings in
+      Lletrec (bindings, traverse ctx body)
```

Comment:
Actually this does not allow destination-passing-style calls in the body of the `let rec` block (because `traverse` is called, not `traverse_binding`) but by reusing the context I believe in the following code:

```ocaml
let[@tailcall_mod_constr] rec f x = fdef in
let[@tailcall_mod_constr] rec g y = gdef in
body
```

you allow destination-passing-style calls to `f` inside `gdef`, which is still a violation of the "only in the defining let rec bindings".

Reply:
Note: `traverse` is actually called for code outside tail position, so this piece of code is responsible of destination-passing-style calls to `f` inside `gdef` if the example code is *not* in tail position.  Destination-passing-style calls to `f` inside *both* `gdef` and `body` are allowed by the similar behavior in `choice` I also pointed out.

See also this [note](https://github.com/ocaml/ocaml/pull/9760#issuecomment-662924359) from @gasche, which questions whether the locality specification is actually what we want, or if the behavior of the current implementation is better (if harder to describe).

Reply:
Yes, currently I believe the description of the PR does not match the implementation, which implements the "global" scope by using a state-monad-style passing for the context.

I am thinking of relaxing the "local scope" specification into an "attribute scope" specification: a TRMC attribute scope over the recursive definitions, and calls are rewritten if they are within the scope of at least one TRMC attribute. This allows the `flatten` example, but it still provides a principled explanation for a strategy that does not require cross-module information for TRMC.

Once this aspect has been decided (I am starting to think that the "attribute scope" proposal is a good compromise, but other people may chime in with suggestions) is will push a separate commit to implement the consensual scoping strategy.